mlflow/types/schema.py

```diff
@@ -408,3 +408,236 @@ def __eq__(self, other) -> bool:
+                f"Unsupported type '{type}', expected instance of DataType or "
+                f"one of {[t.name for t in DataType]}"
+            )
+        if not isinstance(self.type, DataType):
+            raise TypeError(
+                "Expected mlflow.models.signature.Datatype or str for the 'type' "
+                f"argument, but got {self.type.__class__}"
+            )
+
+        # This line makes sure repr(self) works fine
+        self._default = default
+        self._default = self.validate_type_and_shape(repr(self), default, self.type, self.shape)
+
+    @classmethod
+    def validate_param_spec(
+        cls, value: Union[DataType, List[DataType], None], param_spec: "ParamSpec"
+    ):
+        return cls.validate_type_and_shape(
+            repr(param_spec), value, param_spec.type, param_spec.shape
+        )
+
+    @classmethod
+    def validate_type_and_shape(
+        cls,
+        spec: str,
+        value: Union[DataType, List[DataType], None],
+        value_type: DataType,
+        shape: Optional[Tuple[int, ...]],
+    ):
+        """
+        Validate that the value has the expected type and shape.
+        """
+
+        def _is_1d_array(value):
+            if not isinstance(value, list):
+                return False
+            if np.array(value).ndim == 1:
+                return True
+            return False
+
+        try:
+            if shape is None:
+                if np.isscalar(value) or value is None:
+                    return (
+                        np.array(value, dtype=value_type.to_numpy()).item()
+                        if value is not None
+                        else None
+                    )
+                else:
+                    raise MlflowException(
+                        f"Shape must be specified for non-scalar value for ParamSpec {spec}",
+                        INVALID_PARAMETER_VALUE,
+                    )
+            elif shape == (-1,):
+                if _is_1d_array(value):
+                    return (
+                        np.array(value, dtype=value_type.to_numpy()).tolist()
+                        if value is not None
+                        else None
+                    )
+                elif np.isscalar(value):
+                    raise MlflowException(
+                        f"Shape must be None for scalar value for ParamSpec {spec}",
+                        INVALID_PARAMETER_VALUE,
+                    )
+                else:
+                    raise MlflowException(
+                        f"Value must be a scalar or 1D array for ParamSpec {spec}, "
+                        f"received {type(value).__name__}",
+                        INVALID_PARAMETER_VALUE,
+                    )
+            else:
+                raise MlflowException(
+                    "Shape must be None for scalar value or (-1,) for 1D array value "
+                    f"for ParamSpec {spec}), received {shape}",
+                    INVALID_PARAMETER_VALUE,
```

Comment:
I tried to pass in a 2D array and got an exception, which is great:

```
mlflow.exceptions.MlflowException: Invalid parameters found: {('foo', "Value must be a scalar or 1D array for ParamSpec 'foo': string (default: ['baz']) (shape: (-1,)), received list")}
```

However, the exception doesn't tell me the shape of the array that I passed in, which makes it harder to debug. Can we include that info in the exception?